### PR TITLE
feat: add ansible memcached sample

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description of the change:**
Add Memcached Ansible sample

**Motivation for the change:**
operator-framework/enhancements#47

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: hack/generate/samples/ansible/memcached.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+	log.Infof("setting domain and GKV")
+	ma.ctx.Domain = "example.com"
+	ma.ctx.Version = "v1alpha1"
+	ma.ctx.Group = "cache"
+	ma.ctx.Kind = "Memcached"
+}
+
+// Run the steps to create the Memcached Ansible Sample
+func (ma *MemcachedAnsible) Run() {
```
Comment (camilamacedo86) [inline hack/generate/samples/ansible/memcached.go:58]: The Run has all steps to gen the samples. The sample can be checked in the /testdata/ansible
@asmacdo, @jmrodri , @theishshah, @varshaprasad96

File: hack/generate/samples/helm/memcached.go [hunk 2]
```diff
@@ -90,21 +94,16 @@ func (mh *MemcachedHelm) Run() {
...
-	log.Infof("generating OLM bundle")
-	err = mh.ctx.DisableOLMBundleInteractiveMode()
-	pkg.CheckError("generating OLM bundle", err)
-
-	err = mh.ctx.Make("bundle", "IMG="+mh.ctx.ImageName)
-	pkg.CheckError("running make bundle", err)
-
-	err = mh.ctx.Make("bundle-build", "BUNDLE_IMG="+mh.ctx.BundleImageName)
-	pkg.CheckError("running make bundle-build", err)
+	pkg.RunOlmIntegration(mh.ctx)
```
Comment (asmacdo) [inline hack/generate/samples/helm/memcached.go:97]: Since this is out of scope for this work, could you explain the motivation for this refactor?

File: hack/generate/samples/ansible/memcached.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+type MemcachedAnsible struct {
+	ctx *pkg.SampleContext
+}
+
+// NewMemcachedAnsible return a MemcachedAnsible
+func NewMemcachedAnsible(ctx *pkg.SampleContext) MemcachedAnsible {
+	return MemcachedAnsible{ctx}
+}
+
+// Prepare prepare the directory for the Memcached Ansible Sample
```
Comment (asmacdo) [inline hack/generate/samples/ansible/memcached.go:41]:
```suggestion
// Prepare the directory for the Memcached Ansible Sample
```

Comment (asmacdo) [reply hack/generate/samples/ansible/memcached.go:41]: It looks like it does more than just prepare the directory, like setting the Domain and GVK.

File: hack/generate/samples/ansible/memcached.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+func (ma *MemcachedAnsible) Prepare() {
+	log.Infof("destroying directory for memcached Ansible samples")
+	ma.ctx.Destroy()
+
+	log.Infof("creating directory for Ansible Sample")
+	err := ma.ctx.Prepare()
+	pkg.CheckError("error to creating directory for Ansible Sample", err)
```
Comment (asmacdo) [inline hack/generate/samples/ansible/memcached.go:48]:
```suggestion
	pkg.CheckError("error creating directory for Ansible Sample", err)
```

File: hack/generate/samples/ansible/memcached.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+// addingMoleculeMockData will customize the molecule data
+func (ma *MemcachedAnsible) addingMoleculeMockData() {
+	log.Infof("adding molecule test for Ansible task")
+	moleculeTaskPath := filepath.Join(ma.ctx.Dir, "molecule", "default", "tasks",
+		fmt.Sprintf("%s_test.yml", strings.ToLower(ma.ctx.Kind)))
+
+	err := utils.ReplaceInFile(moleculeTaskPath,
+		"false", "true")
+	pkg.CheckError("replacing molecule default tasks", err)
```
Comment (asmacdo) [inline hack/generate/samples/ansible/memcached.go:92]: In the samples repo, I replaced this value with true so we could leave the scaffolded test but make it pass, adding a new file for the new tests. Since you are doing this all in a single file, I don't think it is valuable to edit it in place, why not just template the whole task?

File: hack/generate/samples/ansible/memcached.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+	pkg.CheckError("adding defaulting", err)
+
+	err = utils.ReplaceInFile(filepath.Join(ma.ctx.Dir, "config", "samples",
+		fmt.Sprintf("%s_%s_%s.yaml", ma.ctx.Group, ma.ctx.Version, strings.ToLower(ma.ctx.Kind))),
+		"foo: bar", "size: 1")
+	pkg.CheckError("updating sample CR", err)
+}
+
+// GenerateMemcachedAnsibleSample will call all actions to create the directory and generate the sample
+// Note that it should NOT be called in the e2e tests.
```
Comment (asmacdo) [inline hack/generate/samples/ansible/memcached.go:121]: Could you extend to comment to say why not?

**Review by asmacdo — CHANGES_REQUESTED** [review]
A couple nits and questions.

File: hack/generate/samples/helm/memcached.go (hunk 2 above)
Comment (camilamacedo86) [reply hack/generate/samples/helm/memcached.go:97]: All samples run the same steps to integrate the project with OLM. So, instead of duplicate the code it was centralized in the utils.

File: hack/generate/samples/ansible/memcached.go (hunk 5 above)
Comment (camilamacedo86) [reply hack/generate/samples/ansible/memcached.go:94]: Catcher. Yep, it is true. Good spot.

**Review by asmacdo — APPROVED** [review]

File: hack/generate/samples/generate_all.go [hunk 7]
```diff
@@ -19,33 +19,36 @@ import (
...
-	log.Infof("starting to generate helm memcached sample")
-	ctx, err := pkg.NewSampleContext(binaryName, samplesPath, "GO111MODULE=on")
-	pkg.CheckError("error to generate helm memcached sample", err)
+	log.Infof("creating Helm Memcached Sample")
+	helm.GenerateMemcachedHelmSample(samplesPath)
 
-	log.Infof("creating Memcached Sample")
-	helm.GenerateMemcachedHelmSample(&ctx)
+	log.Infof("creating Ansible Memcached Sample")
+	ansible.GenerateMemcachedAnsibleSample(samplesPath)
```
Comment (camilamacedo86) [inline hack/generate/samples/generate_all.go:53]: The changes here are only to simplify the call of the samples.

File: hack/generate/samples/helm/memcached.go [hunk 8]
```diff
@@ -23,8 +23,10 @@ import (
 
 	"github.com/operator-framework/operator-sdk/hack/generate/samples/pkg"
 	"github.com/operator-framework/operator-sdk/test/utils"
+	testutils "github.com/operator-framework/operator-sdk/test/utils"
```
Comment (camilamacedo86) [inline hack/generate/samples/helm/memcached.go:26]: The changes in this file are only to add better comments, fix log error type msg and keeping all following the same standard.

File: hack/generate/samples/ansible/memcached.go [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+// NewMemcachedAnsible return a MemcachedAnsible
+func NewMemcachedAnsible(ctx *pkg.SampleContext) MemcachedAnsible {
+	return MemcachedAnsible{ctx}
+}
+
+// Prepare the Context for the Memcached Ansible Sample
+// Note that sample directory will be re-created and the context data for the sample
+// will be set such as the domain and GKV.
```
Comment (jmrodri) [inline hack/generate/samples/ansible/memcached.go:43]: nit: It's GVK right?

File: hack/generate/samples/ansible/memcached.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+func (ma *MemcachedAnsible) Prepare() {
+	log.Infof("destroying directory for memcached Ansible samples")
+	ma.ctx.Destroy()
+
+	log.Infof("creating directory")
+	err := ma.ctx.Prepare()
+	pkg.CheckError("creating directory for Ansible Sample", err)
+
+	log.Infof("setting domain and GKV")
```
Comment (jmrodri) [inline hack/generate/samples/ansible/memcached.go:52]: nit: GVK

File: hack/generate/samples/ansible/memcached.go [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+		"foo: bar", "size: 1")
+	pkg.CheckError("updating sample CR", err)
+}
+
+// GenerateMemcachedAnsibleSample will call all actions to create the directory and generate the sample
+// The Context to run the samples are not the same in the e2e test. In this way, note that it should NOT
+// be called in the e2e tests since it will call the Prepare() to set the sample context and generate the files
+// in the testdata directory. The e2e tests only ought to use the Run() method with the TestContext.
+func GenerateMemcachedAnsibleSample(samplesPath string) {
+	ctx, err := pkg.NewSampleContext(testutils.BinaryName, filepath.Join(samplesPath, "ansible/memcached-operator"),
```
Comment (jmrodri) [inline hack/generate/samples/ansible/memcached.go:123]: everywhere else you specified the filenames individually, why are we using the slash (/) as a literal here? shouldn't this be `filepath.Join(samplesPath, "ansible", "memcached-operator"`?

File: hack/generate/samples/ansible/memcached.go [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+		"GO111MODULE=on")
+	pkg.CheckError("generating ansible memcached context", err)
```
Comment (jmrodri) [inline hack/generate/samples/ansible/memcached.go:125]: s/ansible/Ansible

**Review by jmrodri — CHANGES_REQUESTED** [review]
I have one question regarding the `ansible/memcached_operator` string literal being used. Seems like that should be separated. I have a nit, we use `GKV` instead of `GVK` which seems to be more correct.

File: hack/generate/samples/ansible/memcached.go (hunk 11 above)
Comment (camilamacedo86) [reply hack/generate/samples/ansible/memcached.go:123]: Good catcher.

**Comment by camilamacedo86** [comment]
HI @jmrodri, 

All suggestions applied 👍

**Review by jmrodri — APPROVED** [review]
/lgtm
